xdp/src/device.rs

```diff
@@ -0,0 +1,495 @@
+use {
+    crate::{
+        route::Router,
+        umem::{Frame, FrameOffset},
+    },
+    libc::{ifreq, ioctl, mmap, munmap, xdp_ring_offset, IF_NAMESIZE, SIOCGIFADDR, SIOCGIFHWADDR},
+    std::{
+        ffi::{c_char, CStr, CString},
+        io::{self, ErrorKind},
+        marker::PhantomData,
+        mem,
+        net::Ipv4Addr,
+        os::fd::{AsRawFd as _, FromRawFd as _, OwnedFd, RawFd},
+        ptr, slice,
+        sync::atomic::{AtomicU32, Ordering},
+    },
+};
+
+#[derive(Copy, Clone, Debug)]
+pub struct QueueId(pub u64);
+
+pub struct NetworkDevice {
+    if_index: u32,
+    if_name: String,
```

Comment:
Maybe store CString directly?

Reply:
Both ways are not ideal: if we store CString, then we have to convert for NetworkDevice::name(). If we store String, then we have to convert when we do syscalls. Doesn't seem really important either way.